mkdocs/docs/bindings/c-api.md

```diff
@@ -1,9 +1,220 @@
+// the target(s) specified during compilation.
+iree_hal_driver_t* driver = NULL;
+IREE_CHECK_OK(iree_hal_driver_registry_try_create_by_name(
+    iree_hal_driver_registry_default(),
+    iree_string_view_literal("dylib"),
+    iree_allocator_system(), &driver));
+
+// Drivers may support multiple devices, such as when a machine has multiple
+// GPUs. You may either enumerate devices and select based on their properties,
+// or just use the default device.
+iree_hal_device_t* device = NULL;
+IREE_CHECK_OK(iree_hal_driver_create_default_device(
+    driver, iree_allocator_system(), &device));
+
+// Create a HAL module initialized to use the newly created device.
+// We'll load this module into a VM context later.
+iree_vm_module_t* hal_module = NULL;
+IREE_CHECK_OK(
+    iree_hal_module_create(device, iree_allocator_system(), &hal_module));
+// The reference to the driver can be released now.
+iree_hal_driver_release(driver);
+```
+
+!!! tip
+    The default `iree_allocator_system()` is backed by `malloc` and `free`,
+    but custom allocators may also be used.
+
+<!-- TODO(scotttodd): tip about APIs for specific HAL drivers -->
+
+Load a vmfb bytecode module containing program data:
+
+``` c
+// (Application-specific loading into memory, such as loading from a file)
+
+iree_vm_module_t* bytecode_module = NULL;
+IREE_CHECK_OK(iree_vm_bytecode_module_create(
+    iree_const_byte_span_t{module_data, module_size},
+    /*flatbuffer_allocator=*/iree_allocator_null(),
+    /*allocator=*/iree_allocator_system(), &bytecode_module));
+```
+
+!!! note
+    Many IREE samples use
+    [`cc_embed_data`](https://github.com/google/iree/tree/main/build_tools/embed_data)
+    to embed vmfb files as C code to avoid file I/O and ease portability.
+    Applications should use what makes sense for their platforms and deployment
+    configurations.
+
+Create a VM context and load modules into it:
+
+``` c
+iree_vm_context_t* context = NULL;
+iree_vm_module_t* modules[2] = {hal_module, bytecode_module};
+IREE_CHECK_OK(iree_vm_context_create_with_modules(
+    instance, modules, IREE_ARRAYSIZE(modules), iree_allocator_system(),
+    &context));
+// References to the modules can be released now.
+iree_vm_module_release(hal_module);
+iree_vm_module_release(bytecode_module);
+```
+
+Look up the function(s) to call:
+
+``` c
+iree_vm_function_t main_function;
+IREE_CHECK_OK(iree_vm_context_resolve_function(
+    context, iree_string_view_literal("module.main_function"), &main_function));
+```
+
+### Invoke functions
+
+<!-- TODO(scotttodd): I/O buffers (TODO pending new helpers) -->
+<!-- TODO(scotttodd): iree_vm_invoke -->
+
+``` c
+// (Program-specific I/O buffer setup, making data available to the device)
+
+IREE_CHECK_OK(iree_vm_invoke(context, main_function,
+                             /*policy=*/nullptr, inputs, outputs,
+                             iree_allocator_system()));
+
+// (Program-specific output buffer retrieval and reading back from the device)
+```
```

Comment:
I can talk about concepts here (or above) more, but the raw usage in `simple_embedding_test` is pretty verbose: https://github.com/google/iree/blob/59bf64f80f78262ae93a698074097f4e15eb1a8f/iree/samples/simple_embedding/simple_embedding_test.cc#L119-L198